### PR TITLE
convert nxp s32 qspi to zephyr native driver

### PR DESCRIPTION
**Pull request by manuargue**
Convert NXP S32 QSPI controller and flash NOR driver to native drivers. Existing functionalities are retained and some enhancements are done to allow more configuration options. Changes are large, so is better perhaps to review it as a new driver.

```
INFO    - 2764/2778 mr_canhubk3               tests/drivers/flash/common/drivers.flash.common.mr_canhubk3_sfdp_runtime PASSED (device: AB0O4EZU, 11.070s)
INFO    - 2765/2778 mr_canhubk3               samples/subsys/settings/sample.subsys.settings     PASSED (device: AB0O4EZU, 9.580s)
INFO    - 2766/2778 mr_canhubk3               tests/subsys/fs/littlefs/filesystem.littlefs.default PASSED (device: AB0O4EZU, 19.467s)
INFO    - 2767/2778 mr_canhubk3               tests/subsys/storage/flash_map/storage.flash_map.mbedtls PASSED (device: AB0O4EZU, 47.489s)
INFO    - 2768/2778 mr_canhubk3               tests/subsys/logging/log_backend_fs/logging.backend.fs.manualmounted PASSED (device: AB0O4EZU, 18.954s)
INFO    - 2769/2778 mr_canhubk3               tests/subsys/fs/littlefs/filesystem.littlefs.custom PASSED (device: AB0O4EZU, 56.370s)
INFO    - 2770/2778 mr_canhubk3               tests/subsys/settings/functional/fcb/settings.functional.fcb PASSED (device: AB0O4EZU, 12.231s)
INFO    - 2771/2778 mr_canhubk3               tests/subsys/settings/functional/file/settings.file PASSED (device: AB0O4EZU, 14.187s)
INFO    - 2772/2778 mr_canhubk3               tests/drivers/flash/common/drivers.flash.common.default PASSED (device: AB0O4EZU, 9.184s)
INFO    - 2773/2778 mr_canhubk3               tests/subsys/settings/functional/fcb/settings.functional.fcb.chosen PASSED (device: AB0O4EZU, 13.390s)
INFO    - 2774/2778 mr_canhubk3               tests/subsys/settings/fcb/settings.fcb.raw         PASSED (device: AB0O4EZU, 34.381s)
INFO    - 2775/2778 mr_canhubk3               tests/subsys/storage/flash_map/storage.flash_map   PASSED (device: AB0O4EZU, 44.564s)
INFO    - 2776/2778 mr_canhubk3               tests/subsys/fs/fs_api/filesystem.api              PASSED (device: AB0O4EZU, 7.591s)
INFO    - 2777/2778 mr_canhubk3               tests/subsys/storage/flash_map/storage.flash_map.psa PASSED (device: AB0O4EZU, 45.087s)
INFO    - 2778/2778 mr_canhubk3               tests/subsys/settings/file/settings.file.raw       PASSED (device: AB0O4EZU, 9.577s)
```

### CODE REVIEW

**Review by congnguyenhuu — CHANGES_REQUESTED** [review]

File: drivers/memc/memc_nxp_s32_qspi.c [hunk 1]
```diff
@@ -1,197 +1,1707 @@
...
+
+	/* Program DLL configuration */
+	reg_val = REG_READ(QSPI_DLLCR(port));
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_REFCNTR)
+	reg_val &= ~QSPI_DLLCR_DLL_REFCNTR_MASK;
+	reg_val |= QSPI_DLLCR_DLL_REFCNTR(controller->ports[port].dll.reference_counter);
+#endif /* CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_REFCNTR */
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_DLLRES)
+	reg_val &= ~QSPI_DLLCR_DLL_DLLRES_MASK;
+	reg_val |= QSPI_DLLCR_DLL_DLLRES(controller->ports[port].dll.resolution);
```
Comment (congnguyenhuu) [inline drivers/memc/memc_nxp_s32_qspi.c:1055]: these macros `QSPI_DLLCR_DLL_DLLRES_MASK` and `QSPI_DLLCR_DLL_DLLRES` are not defined

File: drivers/memc/memc_nxp_s32_qspi.c [hunk 2]
```diff
@@ -1,197 +1,1707 @@
...
+			},                                                                         \
+		},                                                                                 \
+		.rx_buffers = {                                                                    \
+			.buffers = (struct qspi_buffer *)memc_nxp_qspi_rx_buffers_cfg_##n,         \
+			.buffers_cnt = (sizeof(memc_nxp_qspi_rx_buffers_cfg_##n) /                 \
+					sizeof(struct qspi_buffer)),                               \
+			.all_masters = (bool)DT_INST_PROP(n, ahb_buffers_all_masters),             \
+		},                                                                                 \
+		IF_ENABLED(CONFIG_MEMC_NXP_S32_QSPI_HAS_SFACR_CAS ||                               \
+			   CONFIG_MEMC_NXP_S32_QSPI_HAS_SFACR_BYTE_SWAP, (                         \
```
Comment (congnguyenhuu) [inline drivers/memc/memc_nxp_s32_qspi.c:1675]: you can not use `IF_ENABLED` with a condition that contains the operator. please update.

**Review by congnguyenhuu — CHANGES_REQUESTED** [review]

**Comment by manuargue** [comment]
- addressed all comments above
- split commits for updates in flash driver
- made the vlut for the transfer config of a fixed size to check for boundaries

**Review by bperseghetti — DISMISSED** [review]

**Comment by manuargue** [comment]
- added some parameter validations & assertions
- corrected port base address function for ports > A1

**Comment by manuargue** [comment]
- correct mask for `QSPI_WORD0` that was missed in previous push

**Review by Dat-NguyenDuy — CHANGES_REQUESTED** [review]
Some nits but would like to see addressed

File: drivers/memc/memc_nxp_s32_qspi.c [hunk 3]
```diff
@@ -1,197 +1,1674 @@
...
+	REG_WRITE(QSPI_SMPR, reg_val);
+}
+
+static void qspi_sw_reset(const struct qspi_config *config)
+{
+	volatile uint32_t delay = QSPI_RESET_DELAY_CYCLES;
+
+	/* Software reset AHB domain and Serial Flash domain at the same time */
+	REG_WRITE(QSPI_MCR, REG_READ(QSPI_MCR) | QSPI_MCR_SWRSTHD(1U) | QSPI_MCR_SWRSTSD(1U));
+	delay = QSPI_RESET_DELAY_CYCLES;
```
[the remark this answers is not in the document]
Comment (manuargue) [reply drivers/memc/memc_nxp_s32_qspi.c:948]: I'm using the same criteria as the HAL, in fact this is inherited from it. In the RM is specified in cycles

File: drivers/flash/flash_nxp_s32_qspi_nor.c [hunk 4]
```diff
@@ -615,492 +494,818 @@ static int nxp_s32_qspi_write(const struct device *dev, off_t offset, const void
...
+		.lut_seq = &vlut,
+		.port = config->port,
+		.cmd = NXP_QSPI_READ,
+		.addr = offset,
+		.data = buf,
+		.size = len,
+		.alignment = config->mem_alignment,
+	};
+
+	qspi_release(dev);
```
[the remark this answers is not in the document]
Comment (manuargue) [reply drivers/flash/flash_nxp_s32_qspi_nor.c:776]: yes, fixed

**Comment by manuargue** [comment]
> Is this intended?

not intended, fixed

**Comment by manuargue** [comment]
re-opened

**Comment by manuargue** [comment]
re-opened

File: drivers/flash/flash_nxp_s32_qspi_nor.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dat-NguyenDuy) [reply drivers/flash/flash_nxp_s32_qspi_nor.c:776]: Have you forgot to re-push, i see it has not fixed

[the remark this answers is not in the document]
Comment (manuargue) [reply drivers/flash/flash_nxp_s32_qspi_nor.c]: yes my bad, pushed

**Review by Dat-NguyenDuy — APPROVED** [review]

**Pull request closed without merging**